compiler/rustc_parse/src/parser/item.rs

```diff
@@ -152,15 +152,21 @@ impl<'a> Parser<'a> {
         let lo = self.token.span;
         let vis = self.parse_visibility(FollowedByType::No)?;
         let mut def = self.parse_defaultness();
-        let kind = self.parse_item_kind(
+        let kind = match self.parse_item_kind(
             &mut attrs,
             mac_allowed,
             lo,
             &vis,
             &mut def,
             fn_parse_mode,
             Case::Sensitive,
-        )?;
+        ) {
+            Ok(kind) => kind,
+            Err(err) => {
+                self.sess.parse_errors_encountered.set(true);
```

Comment:
This seems way more flow-dependent than just checking if we've emitted any session errors.

Reply:
For example, this doesn't protect against erroneous recoveries that don't result in a bubbled up `Err` value.

Reply:
Yeah, I experimented with silencing all resolve errors if there were *any* parse errors, but in practice it wasn't an improvement. The only place where the silencing really comes in handy is when parsing patterns fails, and the silencing should be only within that scope, not all resolve errors.